*: Support for both container runtimes, docker and rkt

dongsupark opened the pull request:
Make kube-spawn support not only docker but also rkt. Users are able to run kube-spawn setup with an option `--container-runtime=rkt`.                                                                     

To do that, we change a way of writing most scripts and configs, i.e. generate them according to different k8s version or different container runtimes.

I've only tested with k8s 1.7.5, for both docker and rkt.

Based on patches written by @robertgzr & @iaguis 
Fixes https://github.com/kinvolk/kube-spawn/issues/123

robertgzr commented:
@dongsupark can you squash that fixup commit?

dongsupark commented:
@robertgzr Will do.

nhlfr commented on cmd/kube-spawn/init.go:
```diff
@@ -79,10 +84,37 @@ func doInit() {
+
+	if k8sruntime == "rkt" {
+		kubeadmContainerRuntime = "rktlet"
+	} else {
```
Woudn't it be a little bit more correct if we will look for "docker" value explicitly? Like:

```
switch k8sruntime {
case "", "docker":
    kubeadmContainerRuntime = "docker"
case "rkt":
    kubeadmContainerRuntime = "rkt"
default:
    log.Fatalf("runtime %s is not supported", k8sruntime)
}
```

nhlfr commented on scripts/bootstrap.sh:
```diff
@@ -7,6 +7,7 @@ echo "core:core" | chpasswd
 
 systemctl enable docker.service
 systemctl enable kubelet.service
+systemctl enable rktlet.service
```
Does it mean that we run rktlet unconditionally, even if we are using docker runtine? Or am I missing something?

nhlfr commented on pkg/script/kubeadm-join.go:
```diff
@@ -0,0 +1,25 @@
+
+import "bytes"
+
+const kubeadmJoinTmpl string = `#!/bin/sh
```
I'm a little bit confused. Why you put one script there, in the code, while the another are still in separate files? Can't we template scripts which are in separate files?

dongsupark replied:
@nhlfr Right. Scripts should be basically generated by `pkg/script` to be written to `/var/lib/kube-spawn/scripts`, and scripts under `./scripts` of the source tree, such as bootstrap, init or join, should be removed. They are just leftovers. I'll clean them up.

dongsupark commented on pkg/nspawntool/run.go:
```diff
@@ -152,163 +167,162 @@ func getK8sBindOpts(k8srelease, goPath string) ([]bindOption, error) {
+	case "docker":
+		runtimeDir = "/var/lib/docker"
+	case "rkt":
+		runtimeDir = "/var/lib/rkt"
```
@iaguis Good catch. That's probably why kube-spawn is not able to download large files. Every k8s pods are stored under `/var/lib/rktlet`, not `/var/lib/rkt`. Will fix.

dongsupark commented:
Updated.
* Fixed mount point for rktlet to `/var/lib/rktlet`.
* Added a cmdline option `--rktlet-bin-dir`
* Changed to checking for specific runtime name, `docker` or `rkt`, etc.
* Removed unnecessary scripts and configs from the source tree.

robertgzr commented on cmd/kube-spawn/setup.go:
```diff
@@ -102,65 +112,230 @@ func doSetup(numNodes int, baseImage, kubeSpawnDir string) {
 	}
 
-	log.Printf("All nodes are running. Use machinectl to login/stop/etc.")
+	// NOTE: the 1st part "outbuf" and the 2nd part "outbufRuntime" is merged into
```
why do we merge the unit files here? cri-o and rktlet both use `--container-runtime=remote` and the endpoint option.
Doestn't it make more sense to keep them seperate?

dongsupark replied:
@robertgzr As discussed, I also would like to keep them separate, but it was not that trivial. The thing is, 2 different definitions of `KUBELET_EXTRA_ARGS` need to be merged into 1 definition in the end. Otherwise kubelet won't be able to recognize all the options. So we decided to keep it simply in a single source file.

dongsupark commented:
Updated w.r.t. talks with @robertgzr 
* Combined extra-args and runtime into a single file `kubeadm-extra-args.go`. As the 2 different KUBELET_EXTRA_ARGS definitions need to be combined into a single definition, it doesn't make sense to keep them separate.
* Determined the `failSwapOnArgs` argument before calling `GetKubeadmExtraArgs()`, to avoid tricky syntax in the template.
* Split into multiple commits.

schu commented on pkg/script/kubeadm-bootstrap.go:
```diff
@@ -0,0 +1,27 @@
+
+import "bytes"
+
+const kubeadmBootstrapTmpl string = `#!/bin/sh
```
Why `sh` and not `bash`?

robertgzr replied:
it was that way in the original script... should I change it?

schu commented on pkg/script/kubeadm-bootstrap.go:
```diff
@@ -0,0 +1,27 @@
+systemctl enable kubelet.service
+systemctl enable sshd.service
+
+# necessary to prevent docker from being blocked.
```
Can you explain why / add a link?

schu commented on pkg/script/script.go:
```diff
@@ -0,0 +1,16 @@
+	var buf = new(bytes.Buffer)
+	t := template.Must(template.New(fmt.Sprintf("%T", opts)).Parse(tmpl))
+	if err := t.Execute(buf, opts); err != nil {
+		return nil
```
Why not report the error?

schu commented on cmd/kube-spawn/kube-spawn.go:
```diff
@@ -20,40 +20,81 @@ import (
+
+	cmdKubeSpawn.PersistentPreRun = func(cmd *cobra.Command, args []string) {
+		stderr = log.New(os.Stderr, fmt.Sprintf("[%v] ", cmd.Name()), log.Lshortfile)
+		stdout = log.New(os.Stdout, "", 0)
```
What's the reasoning behind this?

robertgzr replied:
Having info and error output rather than printing everything as a log with date/time

schu replied:
I meant, why do you want some messages go to stdout instead of stderr? Why do messages going to stdout not have the `[cmd]` prefix? If it's just for the format, a `log.SetPrefix(fmt.Sprintf("[%v] ", cmd.Name()))` and `log.SetFlags(log.Lshortfile)` would do as well. What is with all the messages from `pkg/...`?

IMHO, if you want to have "better" logging, try to write down a design first, as far as possible something which suffices both current (again, what about all the code in `pkg`?) and future log usage. And use one of the established libraries in Go (or, if really necessary, build a custom logger pkg compatible with the standard library logger).

Right now it's a wild mix, even in the main package (`git grep 'log\.' cmd/kube-spawn`).

robertgzr replied:
I agree 100%
I started working on the output stuff in a separate commit that should have been excluded from this PR in the first place...

robertgzr replied:
I will open a PR for the logging and output thing.

schu commented on cmd/kube-spawn/setup.go:
```diff
@@ -102,65 +111,220 @@ func doSetup(numNodes int, baseImage, kubeSpawnDir string) {
+	}
+
+	// K8s 1.8 or newer fails to run by default when swap is enabled.
+	// So we should disable the feature with an option "--fail-swap-on=false".
```
Is disabling swap not an option?

schu commented on cmd/kube-spawn/setup.go:
```diff
@@ -102,65 +111,220 @@ func doSetup(numNodes int, baseImage, kubeSpawnDir string) {
+		stdout.Printf("Success! %s started.", node.Name)
+	}
+
+	stdout.Print("All nodes are running. Use machinectl to login/stop/etc.")
```
Nit: "Use machinectl to login ..." is confusing as long as `machinectl login ...` doesn't work :)

Personally, I would prefer better documentation in the README (for example, explain common task such as up, start, stop, delete, etc).

schu commented on pkg/utils/utils.go:
```diff
@@ -135,3 +136,30 @@ func IsTerminal(fd uintptr) bool {
+	return dst, err
+}
+
+func LookupPwd(inPath string) string {
```
Why do we need this function?

robertgzr commented:
Closing in favor of #138
